### DISCUSSION TITLE
pactum global configurations across different requires is a bit confusing

### DISCUSSION THREAD

**Discussion by xusoo** [pactumjs/pactum#115] [Q&A] [answered]
So, I don't know what is happening or how to solve it, neither I've been able to discover the pattern, but take for instance:

- Multiple test file that require a base config file that does `setBaseUrl('URL-A')`
- Other test files that use another base config file that sets a different base URL `'URL-B'`

So depending on which tests run before, the URL A or B is set for the entire test execution, if that makes sense.

Also, not sure if this is related, but something similar happened with the handlers. For some reason, after importing a `base.handlers.js` file into my test file, it wouldn't detect the handlers even if the file did execute (added logs and debugged it). Is the `pactum` instance global or does it depend on how you run/import the tests? Maybe it's worth mentioning that the file was outside the project (like there are multiple sub-projects but all of them inside the same Webstorm project, and I was trying to import this file from one project to another. The require was working, but the handlers weren't being picked up apparently)

Thanks!

**Comment by ASaiAnudeep** [comment, marked as answer]
How does the `setBaseUrl` function is invoked? 

Is it mentioned directly on the root of the file?
```js
// base.config.js
request.setBaseUrl();
```

Or is it wrapped inside a global hook? *Usually this will work as expected*
```js
before(() => {
  request.setBaseUrl();
});

//  tests placeholder
```

> Is the `pactum` instance global or does it depend on how you run/import the tests?

`pactum` maintains a global state. All handlers should register before they are used.

If a sample code is provided to replicate the issues (_a github repo_) , it would help us to investigate it further.

**Reply by xusoo** [reply]
Ok, so apparently wrapping it inside a `before` hook **and requiring the config file inside the top-most `describe` block** makes it work. So that's one thing, thanks!

As for the second part, still couldn't figure out why it doesn't work. Only works if I move it inside my repo.
```js
require('./pactum.handlers'); // Works
require('../../../base-repo/pactum.handlers'); // Doesn't work
```
Is there a reason why it shouldn't work from different NPM projects? I don't get it.

I managed to set up a small repo with two projects where you can reproduce the issue: https://www.dropbox.com/s/f767vc497embe8q/PactumTest.zip?dl=0 

But then, if instead of using the handler directly, I extract it to a utils/helper class (but still in the other base project), it works:

```js
require('./pactum.handlers');

module.exports = {
    async login(email, password) {
        let [cookies, csrfToken] = await pactum.spec('getCookiesAndCsrfToken');
        return pactum.spec('login', { email, password, cookies, csrfToken });
    }
};
```

But to make it even weirder, at the top of this new file, I need to re-set the baseUrl (or require the `base.config.js` again), otherwise, it won't work. This is why I don't completely understand the "global state" of pactum 🤔 

Any clue? 😄

**Reply by ASaiAnudeep** [reply]
This issue is due to the scope of **node_modules**. As you have independent package.json files in two different folders, the node program loads two different pactum instances into the scope. This is not limited to pactum but all other packages behaves the same.

For example, when you load file inside **project-b** from **project-a**, the packages installed under **project-b** will be used.
```sh
project-a
 package.json
project-b
 package.json 
```

To overcome this problem, you can pass the pactum's instance as an argument.

```js
// project-b-handlers.js
module.exports = {
  load(handler) {
    handler.addSpecHandler('testSpec', ctx => {
      ctx.spec.expectError();
    });
  }
}
```

```js
// project-a.test.js
const pactum = require('pactum');
require('path/to/project-b-handlers.js').load(pactum.handler);

describe('Dummy test', () => {
    
    it('should pass', async () => {
        await pactum.spec('testSpec').get('/');
    });
});
```

I am able to load handlers from different projects using the above approach. Hope this helps.

**Reply by xusoo** [reply]
Ok, didn't know about that limitation. That was indeed the workaround I had in mind. Thanks! :)
